// pet-pos-system/src/routes/authRoutes.js
import express from "express";
import bcrypt from "bcryptjs";
import jwt from "jsonwebtoken";

import { pool } from "../../config/db.js";
import { sendOtpEmail } from "../../utils/emailSender.js";

const router = express.Router();

function normalizeEmail(email) {
  return String(email || "").trim().toLowerCase();
}

function generateOtp() {
  return String(Math.floor(100000 + Math.random() * 900000));
}

function createToken(user) {
  if (!process.env.JWT_SECRET) {
    throw new Error("JWT_SECRET is missing.");
  }

  return jwt.sign(
    {
      id: user.id,
      email: user.email,
      role: user.role,
      name: user.name,
    },
    process.env.JWT_SECRET,
    {
      expiresIn: "7d",
    }
  );
}

/**
 * POST /api/auth/login
 */
router.post("/login", async (req, res) => {
  try {
    const email = normalizeEmail(req.body?.email);
    const password = String(req.body?.password || "");

    if (!email || !password) {
      return res.status(400).json({
        ok: false,
        message: "Email and password are required.",
      });
    }

    const result = await pool.query(
      `
      SELECT 
        id,
        name,
        email,
        password_hash,
        role,
        is_active
      FROM users
      WHERE LOWER(email) = LOWER($1)
      LIMIT 1
      `,
      [email]
    );

    const user = result.rows[0];

    if (!user || !user.is_active) {
      return res.status(401).json({
        ok: false,
        message: "Invalid login credentials.",
      });
    }

    const validPassword = await bcrypt.compare(password, user.password_hash);

    if (!validPassword) {
      return res.status(401).json({
        ok: false,
        message: "Invalid login credentials.",
      });
    }

    const token = createToken(user);

    return res.json({
      ok: true,
      message: "Login successful.",
      access_token: token,
      token,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        role: user.role,
      },
    });
  } catch (error) {
    console.error("Login error:", error);

    return res.status(500).json({
      ok: false,
      message: "Login failed.",
    });
  }
});

/**
 * POST /api/auth/forgot-password
 */
router.post("/forgot-password", async (req, res) => {
  try {
    const email = normalizeEmail(req.body?.email);

    if (!email) {
      return res.status(400).json({
        ok: false,
        message: "Email is required.",
      });
    }

    const userResult = await pool.query(
      `
      SELECT 
        id,
        name,
        email,
        is_active
      FROM users
      WHERE LOWER(email) = LOWER($1)
      LIMIT 1
      `,
      [email]
    );

    const user = userResult.rows[0];

    if (!user || !user.is_active) {
      return res.status(404).json({
        ok: false,
        message: "No active admin account found with this email.",
      });
    }

    const otpCode = generateOtp();
    const otpHash = await bcrypt.hash(otpCode, 10);

    await pool.query(
      `
      INSERT INTO password_reset_otps (
        user_id,
        email,
        otp_hash,
        expires_at,
        used_at,
        created_at
      )
      VALUES (
        $1,
        $2,
        $3,
        NOW() + INTERVAL '15 minutes',
        NULL,
        NOW()
      )
      `,
      [user.id, email, otpHash]
    );

    await sendOtpEmail({
      to: email,
      otp: otpCode,
    });

    return res.json({
      ok: true,
      message: "OTP sent to your email.",
    });
  } catch (error) {
    console.error("Forgot password error:", error);

    return res.status(500).json({
      ok: false,
      message: "Failed to generate OTP.",
    });
  }
});

/**
 * POST /api/auth/verify-otp
 */
router.post("/verify-otp", async (req, res) => {
  try {
    const email = normalizeEmail(req.body?.email);
    const otpCode = String(req.body?.otp_code || "").trim();

    if (!email || !otpCode) {
      return res.status(400).json({
        ok: false,
        message: "Email and OTP are required.",
      });
    }

    const result = await pool.query(
      `
      SELECT 
        id,
        user_id,
        email,
        otp_hash,
        expires_at,
        used_at
      FROM password_reset_otps
      WHERE LOWER(email) = LOWER($1)
        AND used_at IS NULL
        AND expires_at > NOW()
      ORDER BY created_at DESC
      LIMIT 1
      `,
      [email]
    );

    const resetOtp = result.rows[0];

    if (!resetOtp) {
      return res.status(400).json({
        ok: false,
        message: "OTP is invalid or expired.",
      });
    }

    const validOtp = await bcrypt.compare(otpCode, resetOtp.otp_hash);

    if (!validOtp) {
      return res.status(400).json({
        ok: false,
        message: "Invalid OTP.",
      });
    }

    return res.json({
      ok: true,
      message: "OTP verified successfully.",
    });
  } catch (error) {
    console.error("Verify OTP error:", error);

    return res.status(500).json({
      ok: false,
      message: "OTP verification failed.",
    });
  }
});

/**
 * POST /api/auth/reset-password
 */
router.post("/reset-password", async (req, res) => {
  const client = await pool.connect();

  try {
    const email = normalizeEmail(req.body?.email);
    const otpCode = String(req.body?.otp_code || "").trim();
    const password = String(req.body?.password || "");
    const confirmPassword = String(req.body?.confirm_password || "");

    if (!email || !otpCode || !password || !confirmPassword) {
      return res.status(400).json({
        ok: false,
        message: "Email, OTP, password and confirm password are required.",
      });
    }

    if (password.length < 8) {
      return res.status(400).json({
        ok: false,
        message: "Password must be at least 8 characters.",
      });
    }

    if (password !== confirmPassword) {
      return res.status(400).json({
        ok: false,
        message: "Passwords do not match.",
      });
    }

    await client.query("BEGIN");

    const userResult = await client.query(
      `
      SELECT 
        id,
        email,
        is_active
      FROM users
      WHERE LOWER(email) = LOWER($1)
      LIMIT 1
      `,
      [email]
    );

    const user = userResult.rows[0];

    if (!user || !user.is_active) {
      await client.query("ROLLBACK");

      return res.status(404).json({
        ok: false,
        message: "No active admin account found with this email.",
      });
    }

    const otpResult = await client.query(
      `
      SELECT 
        id,
        user_id,
        email,
        otp_hash,
        expires_at,
        used_at
      FROM password_reset_otps
      WHERE user_id = $1
        AND LOWER(email) = LOWER($2)
        AND used_at IS NULL
        AND expires_at > NOW()
      ORDER BY created_at DESC
      LIMIT 1
      `,
      [user.id, email]
    );

    const resetOtp = otpResult.rows[0];

    if (!resetOtp) {
      await client.query("ROLLBACK");

      return res.status(400).json({
        ok: false,
        message: "OTP is invalid or expired.",
      });
    }

    const validOtp = await bcrypt.compare(otpCode, resetOtp.otp_hash);

    if (!validOtp) {
      await client.query("ROLLBACK");

      return res.status(400).json({
        ok: false,
        message: "Invalid OTP.",
      });
    }

    const passwordHash = await bcrypt.hash(password, 12);

    await client.query(
      `
      UPDATE users
      SET 
        password_hash = $1,
        updated_at = NOW()
      WHERE id = $2
      `,
      [passwordHash, user.id]
    );

    await client.query(
      `
      UPDATE password_reset_otps
      SET used_at = NOW()
      WHERE id = $1
      `,
      [resetOtp.id]
    );

    await client.query("COMMIT");

    return res.json({
      ok: true,
      message: "Password reset successfully.",
    });
  } catch (error) {
    await client.query("ROLLBACK");

    console.error("Reset password error:", error);

    return res.status(500).json({
      ok: false,
      message: "Failed to reset password.",
    });
  } finally {
    client.release();
  }
});

/**
 * GET /api/auth/me
 */
router.get("/me", async (req, res) => {
  try {
    const authHeader = req.headers.authorization || "";
    const token = authHeader.startsWith("Bearer ")
      ? authHeader.slice(7)
      : null;

    if (!token) {
      return res.status(401).json({
        ok: false,
        message: "Unauthorized. Token is missing.",
      });
    }

    const decoded = jwt.verify(token, process.env.JWT_SECRET);

    const result = await pool.query(
      `
      SELECT 
        id,
        name,
        email,
        role,
        is_active
      FROM users
      WHERE id = $1
      LIMIT 1
      `,
      [decoded.id]
    );

    const user = result.rows[0];

    if (!user || !user.is_active) {
      return res.status(401).json({
        ok: false,
        message: "Unauthorized. User not found or inactive.",
      });
    }

    return res.json({
      ok: true,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        role: user.role,
      },
    });
  } catch (error) {
    console.error("Me error:", error);

    return res.status(401).json({
      ok: false,
      message: "Unauthorized. Invalid token.",
    });
  }
});

export default router;
